### PR TITLE
Enable typed/rackunit source locations.

### PR DESCRIPTION
**Pull request by samth**

### CODE REVIEW

File: rackunit-doc/rackunit/scribblings/misc.scrbl [hunk 1]
```diff
@@ -32,3 +32,27 @@ This example gets @racket[make-failure-test], which is defined in a RackUnit tes
...
+
+@defproc[(check-transformer? [v any/c]) boolean?]{
+Determines if @racket[_v] is a syntax transformer
+defined with @racket[define-check]. Typically, this
+is used on the result of @racket[syntax-local-value].
+
+Provided by @racketmodname[rackunit] at phase 1.
+}
+
+@defproc[(check-transformer-impl-name [ct check-transformer?]) identifier?]{
```
Comment (jackfirth) [inline rackunit-doc/rackunit/scribblings/misc.scrbl]: I'm not sure about this name. Maybe `check-transformer-proc-id`? I'd like the name to clarify that it returns an identifier bound to a plain function. "Impl" and "name" both seem vague to me.

File: rackunit-doc/rackunit/scribblings/misc.scrbl [hunk 2]
```diff
@@ -32,3 +32,27 @@ This example gets @racket[make-failure-test], which is defined in a RackUnit tes
...
+Given a transformer @racket[_ct] defined with @racket[defined-check],
```
Comment (jackfirth) [inline rackunit-doc/rackunit/scribblings/misc.scrbl]: Why the underscore in `@racket[_ct]`?

File: rackunit-doc/rackunit/scribblings/misc.scrbl [hunk 3]
```diff
@@ -32,3 +32,27 @@ This example gets @racket[make-failure-test], which is defined in a RackUnit tes
...
+produces an identifier which names the procedure implementing the
+check. This procedure takes the same arguments as the check form, as
```
Comment (jackfirth) [inline rackunit-doc/rackunit/scribblings/misc.scrbl]: The optional positional string message argument should also be documented here. Personally I think it's a confusing API for a not useful feature, but alas.

File: rackunit-doc/rackunit/scribblings/misc.scrbl [hunk 4]
```diff
@@ -32,3 +32,27 @@ This example gets @racket[make-failure-test], which is defined in a RackUnit tes
...
+well as two mandatory keyword arguments: @racket[#:location] whose argument
+must be a list representing a source location as in the third argument of
```
Comment (jackfirth) [inline rackunit-doc/rackunit/scribblings/misc.scrbl]: Explaining the `#:location` argument should include a reference to `make-check-location`, since they're supposed to take the same kind of value.

File: rackunit-lib/rackunit/private/check.rkt [hunk 5]
```diff
@@ -88,6 +88,22 @@
 
 (define (list/if . vs) (filter values vs))
 
+(begin-for-syntax
```
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check.rkt]: It might be simpler to put this in a separate module to make requiring and providing things at the correct phase easier, but since it's such a small amount of code I'm not sure it's worth it. What do you think?

File: rackunit-lib/rackunit/private/check.rkt [hunk 6]
```diff
@@ -88,6 +88,22 @@
...
+  (require racket/syntax)
+  ;; xform is the actual macro transformer procedure
+  ;; impl-name is an identifier naming the `check-impl` procedure
+  (struct check-transformer (xform impl)
```
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check.rkt]: I like `base` and `proc-id` as names for the struct fields. Not a strong opinion though.

File: rackunit-lib/rackunit/private/check.rkt [hunk 7]
```diff
@@ -88,6 +88,22 @@
...
+    #:property prop:procedure 0)
+  (provide check-transformer-impl-name check-transformer?)
+  (define (check-transformer-impl-name s)
+    (unless (check-transformer? s)
```
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check.rkt]: Why not provide it with a contract?

File: rackunit-lib/rackunit/private/check.rkt [hunk 8]
```diff
@@ -102,22 +118,27 @@
     (with-check-info* infos
       (λ () ((current-check-around) (λ () body ... (void)))))))
 
-(define-simple-macro (define-check (name:id formal:id ...) body:expr ...)
+
```
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check.rkt]: Some extra whitespace slipped in.

File: rackunit-lib/rackunit/private/check.rkt [hunk 9]
```diff
@@ -88,6 +88,22 @@
...
+  ;; impl-name is an identifier naming the `check-impl` procedure
+  (struct check-transformer (xform impl)
+    #:property prop:procedure 0)
+  (provide check-transformer-impl-name check-transformer?)
+  (define (check-transformer-impl-name s)
+    (unless (check-transformer? s)
+      (raise-argument-error
+       'check-transformer-impl-name "check-transformer" s))
+    (check-transformer-impl s))
+  (define-syntax-class check-name
```
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check.rkt]: I think using `id` instead of `name` here and in other places would be clearer and more consistent. Probably should have done that when I refactored `define-check` originally.

File: rackunit-lib/rackunit/private/check.rkt [hunk 10]
```diff
@@ -102,22 +118,27 @@
...
-           #'(lambda args
-               (apply check-impl
-                      #:location (syntax->location #'loc)
-                      #:expression 'chk
-                      args))])))))
+    (define-check-func (name.impl-name formal ...) #:public-name name body ...)
+    (define-syntax name
+      (check-transformer
+       (λ (stx)
+         (with-syntax ([loc (datum->syntax #f 'loc stx)])
```
Comment (jackfirth) [inline rackunit-lib/rackunit/private/check.rkt]: The implementation of this logic in `typed/rackunit` is simpler, and it would be nice to use that instead.

File: rackunit-typed/rackunit/main.rkt [hunk 11]
```diff
@@ -3,66 +3,125 @@
...
-         (for-syntax syntax/parse))
+         (for-syntax syntax/parse syntax/srcloc racket/syntax)
+         (only-in rackunit check-transformer-impl-name))
+
+(begin-for-syntax
+  ;; this implements the behavior of `define-check`, but defers the
+  ;; actual work to `impl-name`
+  ;; impl-name : identifier?
+  ;; stx : syntax?
+  (define (call-impl stx impl-name)
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: Maybe this should go in a (private, undocumented) module shared by `typed/rackunit` and `rackunit`.

File: rackunit-typed/rackunit/main.rkt [hunk 12]
```diff
@@ -3,66 +3,125 @@
...
+         #`(lambda args
+             (apply #,impl-name
+                    #:location 'loc
+                    #:expression 'chk
+                    args))]))))
+
+(define-syntax (define-tcheck stx)
+  (syntax-case stx ()
+    [(_ id)
+     (let* ([impl-id (format-id #'id "~a-imp" #'id)])
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: Using `define-simple-macro` and `#:with impl-id (format-id ...)` would be more readable.

File: rackunit-typed/rackunit/main.rkt [hunk 13]
```diff
@@ -3,66 +3,125 @@
...
+(module impls racket/base
+  (require rackunit (for-syntax racket/base racket/syntax))
+  (define-syntax (get-impls stx)
+    (syntax-case stx ()
+      [(_ id ...)
+       #`(begin
+           #,@(for/list ([id (syntax->list #'(id ...))])
+                (define id-impl (format-id #f "~a-imp" id))
+                (define v (syntax-local-value id #f))
+                (unless (check-transformer? v)
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: Using `define-simple-macro` with a `check-transformer` syntax class that validated and extracted the impl name would be clearer. Maybe that class should even be in rackunit's public API.

File: rackunit-typed/rackunit/main.rkt [hunk 14]
```diff
@@ -3,66 +3,125 @@
...
+                  (raise-syntax-error #f "not a check-transformer" id))
+                #`(begin (provide #,id-impl)
+                         (define #,id-impl
+                           #,(check-transformer-impl-name v)))))]))
+  (get-impls check-equal? check-eq? check-eqv?
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: I like the `provide-impls` name (or something with `provide` in it) better, to make the exporting clearer.

File: rackunit-typed/rackunit/main.rkt [hunk 15]
```diff
@@ -3,66 +3,125 @@
...
+             check-not-equal? check-not-eq? check-not-eqv?
+             check-true check-false check-not-false
+             check-pred check-= check-exn check-not-exn
+             check fail check-regexp-match))
+
+;; this generates the `require/typed` of the impl, the `define-tcheck`
+;; to create the macro wrapper that passes in the source location,
+;; and the provide
+(define-syntax (require-checks stx)
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: The multiple levels of nesting make this macro implementation a little hard to follow. What do you think about adding a `(require/typed/check id t)` helper macro to separate the importing logic from the logic for flattening the checks and type lists?

File: rackunit-typed/rackunit/main.rkt [hunk 16]
```diff
@@ -3,66 +3,125 @@
...
+                               [t (syntax->list #'(t ...))])
+                      (for/list ([id (syntax->list ids)])
+                        (list (format-id id "~a-imp" id) t)))])
+     #'(begin
+         (require/typed 'impls
+                        [id-impl t] ... ...)
+         (define-tcheck id) ... ...
+         (provide id ... ...)))]))
+;; type for things like `check-equal?`
+(define-type check-impl-ish-ty
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: Some duplication could be removed with a `(define-check-type type-id arg-type ...)` macro that expanded to `(define-type type-id (case-> (arg-type ... #:loc ... #:expr ...) (arg-type ... String #:loc ... #:expr ...))`. Might be worth supporting the polymorphic check types too.

File: rackunit-typed/rackunit/main.rkt [hunk 17]
```diff
@@ -3,66 +3,125 @@
...
+                   ((Thunk Any) #:location Any #:expression Any -> Any)
+                   ((Thunk Any) String #:location Any #:expression Any -> Any))]
+                [(check)
+                 (All (A B)
+                      (case->
+                        ((A B -> Any) A B #:location Any #:expression Any -> Any)
+                        ((A B -> Any) A B String #:location Any #:expression Any -> Any)))]
+                [(fail)
+                 (->* (#:location Any #:expression Any) (String) Void)]
+                [(check-regexp-match)
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: Huh, didn't realize until now that `check-regexp-match` doesn't accept a message argument. There should probably be an issue opened for that in the name of consistency.

File: rackunit-typed/rackunit/main.rkt [hunk 18]
```diff
@@ -3,66 +3,125 @@
...
+;; various checks we need, so that we have something to point
+;; to with `require/typed`
+(module impls racket/base
+  (require rackunit (for-syntax racket/base racket/syntax))
+  (define-syntax (get-impls stx)
+    (syntax-case stx ()
+      [(_ id ...)
+       #`(begin
+           #,@(for/list ([id (syntax->list #'(id ...))])
+                (define id-impl (format-id #f "~a-imp" id))
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: Would this work without the extra definition? As in, just expanding to `(provide (rename-out [<impl-id> <imp-id>] ...))`.

File: rackunit-typed/rackunit/main.rkt [hunk 19]
```diff
@@ -3,66 +3,125 @@
...
+    (Any Any String #:location Any #:expression Any -> Any)))
+
+;; type for things like `check-true`
+(define-type unary-check-impl-ish-ty
+  (case->
+    (Any #:location Any #:expression Any -> Any)
+    (Any String #:location Any #:expression Any -> Any)))
+
+;; all the actual specifications
+(require-checks [(check-equal? check-eq? check-eqv?
```
Comment (jackfirth) [inline rackunit-typed/rackunit/main.rkt]: Starting the clauses on the next line would make it easier to read and format them.

**Comment by jessealama** [comment]
Is there any traction on this? I'm getting stung (or, rather, beaten up) by the loss of source locations in TR's rackunit.

**Comment by jessealama** [comment]
any traction here?

**Comment by samth** [comment]
If you want to work on this, I am happy to defer to you. I also should have time to pick this up if you aren't going to get to it soon.

I believe the real issue is that I think parts of this didn't work -- ie, broke running some test cases, but I don't remember any details.

**Comment by jessealama** [comment]
@samth I probably won't be able to get to this terribly soon; I just wanted to offer help, if needed.

Looks like @bennn suggested something in a racket-users thread (see [here](https://github.com/racket/rackunit/issues/120)).

**Comment by capfredf** [comment]
I was bitten by this issue again, so let me take this over.

**Comment by samth** [comment]
Wasn't this superseded by #153 ?

**Comment by jbclements** [comment]
Yes, I believe it was.
